### ISSUE TITLE
Mouse movement on mac with dual displays

### ISSUE THREAD

**Issue by GuruGurra** [TolikPylypchuk/SharpHook#39]
I have a program that presses the mouse button, moves the mouse and then releases the button. The movement is horizontal only, so the Y parameter is always the current position.

On a mac with a single display, it works as expected.
On a mac with two displays, the cursor is moved up 72 steps for each call, even though I tell it not to move vertically.
Both are using Monterey.

Is the dual display causing this problem, or can it be something else? If it is a dual display problem, how do I solve it?

**Comment by Berets** [comment]
Possible duplicate of #36

**Comment by TolikPylypchuk** [comment]
Hi! Thanks for asking!

Unfortunately, I don't really know what causes the issue or how to help you. Also, I cannot check this behavior, because I don't have a Mac.

This seems like an issue in libuiohook - the underlying C library. Could you please create an issue in the [original libuiohook repo](https://github.com/kwhat/libuiohook) instead? Though I should warn you that it's currently not under active development.

I'm going to close this issue, but we can continue the discussion here if you wish.

**Comment by GuruGurra** [comment]
I've realized that this maybe isn't a multi-display problem but something else.

To explain the setup: I write a Stream Deck plugin. The plugin is run as a separate process, started by the Stream Deck engine. When the process starts, I execute the following sequence:

hook = new TaskPoolGlobalHook();
hookTask = hook.RunAsync();
hook.MouseMoved += Hook_MouseMoved;
hook.MouseDragged += Hook_MouseDragged;
hook.HookEnabled += Hook_HookEnabled;
hook.HookDisabled += Hook_HookDisabled;

(by the way - the Global Hooks page in the doc says, "You can subscribe to events after the hook is started.", but the example code subscribes to events before it is started...unless you with "started" mean "created", not "run").

Anyhow, I subscribe to mouse-moved/dragged events and receive callbacks as expected. This is until the Stream Deck editor for the plugin is opened, at which point, all activities from SharpHook seem to stop.

I also subscribe to log events with minLevel: LogLevel.Debug. Before the opening of the Stream Deck editor, the normal flow of events looks like this:

[process_mouse_moved [831]: Mouse moved to 365, 889.]
[dispatch_event [125]: Dispatching event type 9.]
[process_mouse_moved [831]: Mouse moved to 365, 889.]
[dispatch_event [125]: Dispatching event type 9.]
[process_mouse_moved [831]: Mouse moved to 364, 889.]
[dispatch_event [125]: Dispatching event type 9.]
[process_mouse_moved [831]: Mouse moved to 363, 889.]
[dispatch_event [125]: Dispatching event type 9.]


When I open the Stream Deck editor, I get the following sequence:

[process_mouse_moved [831]: Mouse moved to 361, 891.]
[dispatch_event [125]: Dispatching event type 9.]
[process_mouse_moved [831]: Mouse moved to 361, 891.]
[dispatch_event [125]: Dispatching event type 9.]
[process_mouse_moved [831]: Mouse moved to 361, 891.]
[dispatch_event [125]: Dispatching event type 9.]
[process_system_key [538]: Using dispatch_sync_f for system key events.]

After that, I don't get any more callbacks from SharpHook, nothing.

hook.IsRunning == true, and I get no Hook_HookDisabled callback. I can still call the simulator to move the cursor, but gets no callback whatsoever from SharpHook.

What is going on?

By the way - This is on mac Ventura, but not on all computers, it may be limited to the M1 CPU, but that's a wild guess. I have a mac M2 Ventura, and I have seen the problem once on my mac but cannot reproduce it. Other users have this behavior constantly.

Best,
Gunnar

________________________________
Från: Tolik Pylypchuk ***@***.***>
Skickat: den 8 mars 2023 22:03
Till: TolikPylypchuk/SharpHook ***@***.***>
Kopia: Gunnar Carlson ***@***.***>; Author ***@***.***>
Ämne: Re: [TolikPylypchuk/SharpHook] Mouse movement on mac with dual displays (Issue #39)


Hi! Thanks for asking!

Unfortunately, I don't really know what causes the issue or how to help you. Also, I cannot check this behavior, because I don't have a Mac.

This seems like an issue in libuiohook - the underlying C library. Could you please create an issue in the original libuiohook repo<https://github.com/kwhat/libuiohook> instead? Though I should warn you that it's currently not under active development.

I'm going to close this issue, but we can continue the discussion here if you wish.

**Comment by TolikPylypchuk** [comment]
Hm, this is definitely weird. I still think it's a libuiohook issue, but I'm not 100% sure, so I've reopened this issue.

As for subscribing to events - you can subscribe to them after the hook is _started_. In the example they are subscribed to before starting, and that's why I've added an explanation. An exception to this is `HookEnabled` as it's raised once when you call `Run` or `RunAsync` and never again, so subscribing to it afterwards will have no effect.

**Comment by GuruGurra** [comment]
Is there anything I can do to gather more information?
________________________________
Från: Tolik Pylypchuk ***@***.***>
Skickat: den 10 mars 2023 23:03
Till: TolikPylypchuk/SharpHook ***@***.***>
Kopia: Gunnar Carlson ***@***.***>; Author ***@***.***>
Ämne: Re: [TolikPylypchuk/SharpHook] Mouse movement on mac with dual displays (Issue #39)


Hm, this is definitely weird. I still think it's a libuiohook issue, but I'm not 100% sure, so I've reopened this issue.

As for subscribing to events - you can subscribe to them after the hook is started. In the example they are subscribed to before starting, and that's why I've added an explanation. An exception to this is HookEnabled as it's raised once when you call Run or RunAsync and never again, so subscribing to it afterwards will have no effect.

**Comment by TolikPylypchuk** [comment]
I don't think so. As I've said, the problem is that I don't have a Mac so I can't check this issue at the moment.

**Comment by GuruGurra** [comment]
I have realized that the "process_system_key [538]: Using dispatch_sync_f for system key events." message is sent when the (physical) mouse button is pressed.

After receiving the "process_system_key [538]: Using dispatch_sync_f for system key events." message", I experimented with disposing of the hook and recreating it. This had the following consequences:

1. When I dispose of the hook, I get a "hook stopped" message but no HookDisabled event.
2. After recreating the hook, events and messages are received as normal again.
3. After recreating the hook, the mouse buttons (on the physical mouse) are inoperable. I can move the mouse around but get no reaction when using mouse buttons. 

So...
- If I do nothing after having received that message, I lose all future events and messages.
- If I restart the hook after having received that message, the physical mouse buttons stop working.

**Comment by TolikPylypchuk** [comment]
Hm, this sounds like a weird macOS issue or libuiohook not handling mouse events for a physical mouse correctly.

**Comment by TolikPylypchuk** [comment]
Hello @GuruGurra. I finally have access to a Mac, so I've tried to reproduce your issues, and I couldn't.

Regarding the mouse movement simulation and vertical jumps - a new way to simulate mouse movement was added since we discussed this issue. You can now use `SimulateMouseMovementRelative(x, y)` in order to move the mouse by an offset relatvie to the current cursor position. Can you please try this approach (with `y = 0`) and tell me whether the issue is reproduced?

Regarding the second issue - I don't have a Stream Deck, so I can't know what exactly is going to. One thing to note here - `Using dispatch_sync_f for system key events` is logged when the global hook is not run on the main thread. macOS requires that the main thread have a run-loop present, or else the events won't be captured. Maybe, Stream Deck editor interferes with it somehow. Also, I couldn't reproduce the physical mouse issues with disposing a hook and creating a new hook. If you could reproduce these issues in a plain console app, that would be great.

**Comment by TolikPylypchuk** [comment]
I've just released version 5.0.0 and while it doesn't directly fix your issues, it did change a lot of stuff internally, so please try reproducing the problems using that version. I'm going to close this issue for now, but if the problems are reproduced again, I'll reopen this issue and we can continue the discussion.
